### PR TITLE
State

### PR DESCRIPTION
**Pull request by mantono**

### CODE REVIEW

File: src/main/kotlin/JobState.kt [hunk 1]
```diff
@@ -0,0 +1,64 @@
+package franz.internal
```
Comment (williamhogman) [inline src/main/kotlin/JobState.kt]: If its a public API it should live in package Franz

File: src/main/kotlin/JobState.kt [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+fun <T, U> JobDSL<T, U>.asPipe(): JobState<U> = JobState(this.value)
+
+class JobState<U>(val value: U){
+    var status: JobStatus = JobStatus.Incomplete
+
+        get() = field
+
+        set(value) {
+            if(status == JobStatus.Incomplete)
```
Comment (williamhogman) [inline src/main/kotlin/JobState.kt:14]: This must throw, also shouldn't this be private?

Comment (williamhogman) [reply src/main/kotlin/JobState.kt:14]: Also why is this setable at at all. All calls are copy constructors.

File: src/main/kotlin/JobState.kt [hunk 3]
```diff
@@ -0,0 +1,64 @@
+package franz.internal
+
+fun <T, U> JobDSL<T, U>.asPipe(): JobState<U> = JobState(this.value)
+
+class JobState<U>(val value: U){
```
Comment (williamhogman) [inline src/main/kotlin/JobState.kt]: Constructor should be `internal`

File: src/main/kotlin/JobState.kt [hunk 4]
```diff
@@ -0,0 +1,64 @@
+package franz.internal
+
+fun <T, U> JobDSL<T, U>.asPipe(): JobState<U> = JobState(this.value)
```
Comment (williamhogman) [inline src/main/kotlin/JobState.kt]: If we refer to the concept as a pipe in JobDSL then our class should be called as pipe.

2. If we want to keep JobDSL.Kt clean from dependencies on pipe then the pipe stuff should have its own namespace.

File: src/main/kotlin/JobState.kt [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+     * Use when conducting the final operation on the job. If it successfully done (the predicate returns true)
+     * the [JobState] will automatically be set as [JobStatus.Success] and return the [JobStatus] rather than
+     * the [JobState] itself, prohibiting that any more work is done on this job through this pipe. The returned
+     * [JobStatus] can never be [JobStatus.Incomplete] when returning from this function.
+     * */
+
+    fun end(predicate: (U) -> Boolean): JobStatus {
+        val state = process(JobStatus.TransientFailure, predicate)
+        if(state.inProgress())
+            state.status = JobStatus.Success
```
Comment (williamhogman) [inline src/main/kotlin/JobState.kt]: Copy constructor instead?

File: src/main/kotlin/JobState.kt [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        return state.status
+    }
+
+    private inline fun process(newStatus: JobStatus, predicate: (U) -> Boolean): JobState<U> {
+        if(inProgress() && !predicate(value))
+            this.status = newStatus
+        return this
```
Comment (williamhogman) [inline src/main/kotlin/JobState.kt]: copy constructor

File: src/main/kotlin/JobState.kt [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    fun <R> map(transform: (U) -> R): JobState<R> {
+        if(!inProgress())
```
Comment (williamhogman) [inline src/main/kotlin/JobState.kt]: Why does map fail when process doesn't? I agree that we want an API with this failure semantic, though.

**Review by williamhogman — CHANGES_REQUESTED** [review]
Fix/Clarify based on comments :)

File: src/main/kotlin/JobState.kt (hunk 1 above)
Comment (mantono) [reply src/main/kotlin/JobState.kt]: Oh, I only found _franz.internal_ so I used that too, but it felt wrong to do so. I'll fix that.

File: src/main/kotlin/JobState.kt (hunk 2 above)
Comment (mantono) [reply src/main/kotlin/JobState.kt:14]: 1. What should throw? The getter? It cannot be private, because then the _status_ property cannot be accessed from outside the class.

2. Because if we do
```
val state: JobState = job.asPipe()
    .validate { doSomethingThatWorks }
    .validate { doSomethingThatWorks }

(some intermediary operation)

state
    .validate { doSomethingThatWorks }
    .validate { doSomethingThatFails }

(some intermediary operation again)

state
    .validate { doSomething } // <-- this should not be possible, but it is if state is only passed 
// down through chaining
```

Currently, no calls are copy constructor?

File: src/main/kotlin/JobState.kt (hunk 5 above)
Comment (mantono) [reply src/main/kotlin/JobState.kt]: Function _end_ should return a JobStatus, so it would be very strange if it returned another JobState.

File: src/main/kotlin/JobState.kt (hunk 6 above)
Comment (mantono) [reply src/main/kotlin/JobState.kt]: See https://github.com/zensum/franz/pull/4#discussion_r131153045

File: src/main/kotlin/JobState.kt (hunk 7 above)
Comment (mantono) [reply src/main/kotlin/JobState.kt]: Because you're trying to perform an action on a job that is no longer possible to do something with. I should probably update the docs to reflect this.

An example of such a scenario is when prior to doing the map operation, a job has already been given the status _Success_ because it was already found in an idempotence store and should no longer be worked with or it has a _PermanentFailure_ status because some previous operation returned an error. The fact that this function throws an exception intends to communicate that this is not normal program flow and should only happen when you miss to check that all the necessary prerequisites have been met. 

So in other words, it does fail when **a** process fails, but a prior process and not the current one.

File: src/main/kotlin/JobState.kt (hunk 3 above)
Comment (mantono) [reply src/main/kotlin/JobState.kt]: Good point!

File: src/main/kotlin/JobState.kt (hunk 5 above)
Comment (williamhogman) [reply src/main/kotlin/JobState.kt]: Ok what about a big-If expression?

Comment (mantono) [reply src/main/kotlin/JobState.kt]: I am not really sure what you mean with a big if expression.

Comment (williamhogman) [reply src/main/kotlin/JobState.kt]:
```if(state.inProgress()) {
  return JobStatus.Success
}
```
and not mutate the thing.

**Review by williamhogman — APPROVED** [review]
